src/Amadevus.RecordGenerator.Generators/EqualityUsageAnalyzer.cs

```diff
@@ -0,0 +1,28 @@
+﻿using Amadevus.RecordGenerator.Analyzers;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Amadevus.RecordGenerator.Generators
+{
+    internal sealed class EqualityUsageAnalyzer
+    {
+        public static IEnumerable<Diagnostic> GenerateDiagnostics(RecordDescriptor descriptor)
+        {
+            if (!descriptor.Features.HasFlag(Features.EquatableEquals) 
+             && !descriptor.Features.HasFlag(Features.ObjectEquals)
+             && !descriptor.Features.HasFlag(Features.OperatorEquals)
+             || descriptor.TypeDeclaration.Modifiers.Any(t => t.IsKind(SyntaxKind.SealedKeyword)))
+            {
+                yield break;
+            }
+
+            yield return Diagnostic.Create(
+                Descriptors.X1001_RecordMustBeSealedIfEqualityIsEnabled,
+                descriptor.TypeDeclarationLocation,
+                descriptor.TypeIdentifier.Text);
+        }
+        
+    }
```

Comment:
In the end, I think we may want to keep analysis in the Analyzers project anyway. There we can offer a CodeFix that'll make the class sealed, enhancing the experience.

Reply:
Moved to #76